packages/fetchai/connections/soef/connection.py

```diff
@@ -91,106 +147,349 @@ def __init__(
+
+            assert isinstance(envelope.message, OefSearchMessage), ValueError(
+                "Message not of type OefSearchMessage"
             )
-            self.search_services(self.search_id, query)
-        else:
-            raise ValueError("OEF request not recognized.")
 
-    def register_service(self, service_description: Description) -> None:
+            oef_message = cast(OefSearchMessage, envelope.message)
+
+            handlers_and_errors = {
+                OefSearchMessage.Performative.REGISTER_SERVICE: (
+                    self.register_service,
+                    err_ops.REGISTER_SERVICE,
+                ),
+                OefSearchMessage.Performative.UNREGISTER_SERVICE: (
+                    self.unregister_service,
+                    err_ops.UNREGISTER_SERVICE,
+                ),
+                OefSearchMessage.Performative.SEARCH_SERVICES: (
+                    self.search_services,
+                    err_ops.SEARCH_SERVICES,
+                ),
+            }
+
+            if oef_message.performative not in handlers_and_errors:
+                raise ValueError("OEF request not recognized.")
+
+            handler, oef_error_operation = handlers_and_errors[oef_message.performative]
+            await handler(oef_message)
+
+        except SOEFException:
+            await self._send_error_response(oef_error_operation=oef_error_operation)
+        except Exception:  # pylint: disable=broad-except
+            logger.exception("Exception during envelope processing")
+            await self._send_error_response(oef_error_operation=oef_error_operation)
+            raise
+
+    async def register_service(self, oef_message: OefSearchMessage) -> None:
         """
         Register a service on the SOEF.
 
-        :param service_description: the service description
-        """
-        if self._is_compatible_description(service_description):
-            service_location = service_description.values.get("location", None)
-            piece = service_description.values.get("piece", None)
-            value = service_description.values.get("value", None)
-            if service_location is not None and isinstance(service_location, Location):
-                self._set_location(service_location)
-            elif isinstance(piece, str) and isinstance(value, str):
-                self._set_personality_piece(piece, value)
-            else:
-                self._send_error_response()
-        else:
-            logger.warning(
-                "Service description incompatible with SOEF: values={}".format(
-                    service_description.values
-                )
+        :param oef_message: OefSearchMessage
+        """
+        service_description = oef_message.service_description
+
+        data_model_handlers = {
+            "location_agent": self._register_location_handler,
+            "personality_agent": self._register_personality_piece_handler,
```

Comment:
I think it should be called `set_personality_piece`

Reply:
updated